CRM/Financial/BAO/PaymentProcessor.php

```diff
@@ -569,6 +569,30 @@ public static function getPaymentProcessorForRecurringContribution($contribution
     return Civi\Payment\System::singleton()->getById($paymentProcessorId);
   }
 
+  /**
+   * Get the name of the payment processor
+   *
+   * @param $paymentProcessorId
+   *
+   * @return null|string
+   */
+  public static function getPaymentProcessorName($paymentProcessorId) {
```

Comment:
so if the BAO calls getting replaced by API calls elsewhere above, why to create new BAO method (even if it at the end use the API) instead of calling the API directly, because to be honest I am not clear on the line between when to use the API directly and when to use BAO methods in civicrm core code.

Reply:
@omarabuhussein @eileenmcnaughton Good question!  I wanted to to something consistent (it's possible to have a recurring contribution with no payment processor ID, or an invalid one).  I couldn't find a way of doing this via the UI without duplicating around 6 lines of code (which I abstracted into getPaymentProcessorName).  Any thoughts?

Reply:
I guess my position is that it's good to use the api because it's standardised & various hooks (include api-layer) work if you use the api but when there is a re-usable chunk then it makes sense to wrap that in a function. In this case the handling of the unknown situation is an extra bit of handling that the function provides. I would have thought that the FKs on the contribution_recur table would have meant we should not see an invalid processor id - but I'm not opposed to handling the posibility.

CiviCRM is never going to be a purist's codebase :-) which does mean we can't rely on rules as easily as we might in other codebases.

Reply:
okay then :)